posts/2018-06-15-typedapi.md

```diff
@@ -0,0 +1,329 @@
+GET /users/:name?minAge=:age -> List[User]
+```
+
+It only consists of a single endpoint which returns a list of `Users`:
+
+```scala
+final case class User(name: String, age: Int)
+```
+
+with a given `name: String`. Furthermore, you filter the resulting users by their `age: Int`. Our big goal is to end up with a function which is derived from a type-level representation of our endpoint:
+
+```scala
+(name: String, minAge: Int) => F[List[User]]
+```
+
+### Represent the api as a type
+Question: how do you represent the above api as a type in Scala? I think the best way is to break it apart and try to find type-level representations for each element. After that, we "just" merge them together.
+
+When we take a closer look at our endpoint we see that it consists of:
+  * path elements identifying an endpoint: `/users`
+  * segments within the path: `:name`
+  * queries: `minAge=[age]`
+  * and a method `GET` to identify which kind of operation we want to do
+
+Or in other words, just a plain HTTP definition of a web endpoint. Now that we know what we are working with let's try and find a type-level representation.
+
+We start with our path element. If we would work with Dotty or Typelevel-Scala we could use a feature called literal-type:
+
+```scala
+type Path = "users"
+```
+
+But as we want to stay in Vanilla Scala this isn't an option.
+We have to use the one tool probably every developer has to use when it comes to working on the type-level called [shapeless](https://github.com/milessabine/shapeless).
+It has this nifty class [Witness](https://github.com/milessabin/shapeless/blob/shapeless-2.3.3/core/src/main/scala/shapeless/singletons.scala#L32) which generates a witness type from a literal like `Strings` or `Symbols`.
+
+```scala
+import shapeless.Witness
+
+val usersW = Witness("users")
+```
+
+But this isn't a pure type declaration, you will say. And you are right, but right now there is no other way in Scala. We have to go the ordinary value road first to create our types.
+
+Now that we know how to get a type representation from a `String` which describes our path we should clearly mark it as a path element:
+
+```scala
+sealed trait Path[P]
+
+type users = Path[usersW.T]
+```
+
+That's it. That is the basic concept of how we can describe our apis as types. We just reuse this concept now for the remaining elements like the segment.
+
+
+```scala
+val nameW = Witness('name)
+
+sealed trait Segment[K, V]
+
+type name = Segment[nameW.T, String]
+```
+
+Do you see how we included the segment's identifier in the type? This way we are not only gain information about the expected type but also what kind of value we want to see. By the way, I decided to use `Symbols` as identifiers, but you could also switch to `String` literals. The remaining definitions look pretty similar:
+
+```scala
+val minAgeW = Witness('minAge)
+
+sealed trait Query[K, V]
+
+type minAge = Query[minAgeW.T, Int]
+
+sealed trait Method
+sealed trait Get[A] extends Method
```

Comment:
It's worth mentioning that the `A` type parameter will represent the type of the resource returned by the endpoint.